src/NuGet.Core/NuGet.Configuration/Settings/Items/SourceItem.cs

```diff
@@ -77,8 +77,7 @@ public override bool Equals(object other)
                 return true;
             }
 
-            return string.Equals(Key, source.Key, StringComparison.Ordinal) &&
-                string.Equals(ProtocolVersion, source.ProtocolVersion, StringComparison.Ordinal);
+            return string.Equals(Key, source.Key, StringComparison.Ordinal);
```

Comment:
Why are we making change here? I believe it should stay as it's now.

Reply:
This change is needed as otherwise using `nuget sources update` with protocol version specified results in adding new item to sources list with the same name as already existing instead of updating.

This additionally fixes a bug that is there right now where if you use `nuget sources update` on a source that has protocol version specified, e.g. nuget v3 feed, this operation is resulting in nuget config with two sources with the same name.

Reply:
Thank you for explanation, sounds reasonable for me. What others think?

Reply:
Team Review: We think it would best if the code path that needs a custom comparison implements it that way rather than changing how all comparisons work.  This could be a breaking change at the moment to alter how these objects are compared.

Reply:
Several tests are failing, please check why they're failing: [#1](https://github.com/NuGet/NuGet.Client/blob/9836b4469a575b6b1aa7c9072c28b2f4ef6ec91d/test/NuGet.Core.Tests/NuGet.Configuration.Test/SettingsFileParsingTests/SourceItemTests.cs#L128), [#2](https://github.com/NuGet/NuGet.Client/blob/9836b4469a575b6b1aa7c9072c28b2f4ef6ec91d/test/NuGet.Core.Tests/NuGet.Configuration.Test/PackageSourceProviderTests.cs#L1029), [#3](https://github.com/NuGet/NuGet.Client/blob/9836b4469a575b6b1aa7c9072c28b2f4ef6ec91d/test/NuGet.Core.Tests/NuGet.Configuration.Test/PackageSourceProviderTests.cs#L1624).

Reply:
Thank you for comments and running tests.

@erdembayar, the tests are failing due to aforementioned change in the comparison. My bad for missing that those tests failed. I'll revert the change.

@jeffkl 
What should be the expected behaviour in the below scenario?


Source item in `nuget.config` before change: `<add key="Source-Name" value="Legacy-Source" />`
Command: `nuget sources update -Name "Source-Name" -Source "New-Source" -ProtocolVersion 3`

Looking at [#2](https://github.com/NuGet/NuGet.Client/blob/9836b4469a575b6b1aa7c9072c28b2f4ef6ec91d/test/NuGet.Core.Tests/NuGet.Configuration.Test/PackageSourceProviderTests.cs#L1029) seems that the above command should result in `nuget.config` ending up with two lines for `Source-Name`:
- `<add key="Source-Name" value="Legacy-Source" />`
- `<add key="Source-Name" value="New-Source" protocolVersion="3" />`

Please confirm if my understanding is correct or different behaviour is expected.
I will then push the revert for comparison and updated tests.

Reply:
If I understand correctly Team triage [comment](https://github.com/NuGet/NuGet.Client/pull/5002#discussion_r1072625616) asks you revert this comparison logic change.  I understand your intention, but it might break other scenarios we're not aware of. Once this's reverted then I believe failing tests would be passing.

Reply:
I have pushed the revert for comparison logic. Previously failing tests [#1](https://github.com/NuGet/NuGet.Client/blob/9836b4469a575b6b1aa7c9072c28b2f4ef6ec91d/test/NuGet.Core.Tests/NuGet.Configuration.Test/SettingsFileParsingTests/SourceItemTests.cs#L128), [#2](https://github.com/NuGet/NuGet.Client/blob/9836b4469a575b6b1aa7c9072c28b2f4ef6ec91d/test/NuGet.Core.Tests/NuGet.Configuration.Test/PackageSourceProviderTests.cs#L1029), [#3](https://github.com/NuGet/NuGet.Client/blob/9836b4469a575b6b1aa7c9072c28b2f4ef6ec91d/test/NuGet.Core.Tests/NuGet.Configuration.Test/PackageSourceProviderTests.cs#L1624) succeeded when running locally.

That however causes change in behaviour of `dotnet sources update` command when `-protocolVersion` is specified and is different than the value in nuget.config. That, in turn, causes tests added for new option to fail.

While taking closer look at failed tests mentioned by you, I am not sure if the behaviour of `sources update` command I implemented for `-ProtocolVersion` is the correct one since it is inconsistent with other commands (e.g. in test [#2](https://github.com/NuGet/NuGet.Client/blob/9836b4469a575b6b1aa7c9072c28b2f4ef6ec91d/test/NuGet.Core.Tests/NuGet.Configuration.Test/PackageSourceProviderTests.cs#L1029)).
That's the reason for [my previous comment](https://github.com/NuGet/NuGet.Client/pull/5002#discussion_r1080616770).

Without the comparison logic change, the code behaves as in scenario described. `nuget.config` has two lines for the source with the same name and different values and protocol versions.
If that is the expected behaviour, I need to fix tests for sources command with Protocol Version specified. They are failing right now, expecting only one line for source in `nuget.config` after sources update command.

Alternatively, I need to amend the behaviour of the `PackageSourceProvider.UpdatePackageSource` for this case. That does seem quite cumbersome without the comparison logic change, so any suggestions and pointers are welcome. 

With that said, I am happy to go either way with the implementation.

I hope that better explains my intent 😃

Reply:
I've pushed a change that retains earlier behaviour from this PR but does not require comparison logic change.

Please let me know if that is directionally correct.

Reply:
Thank you for pushing new change. I just requeued the CI build, let's see how it goes.

Reply:
Some tests failed, could be flaky test, I just requeued the failed tests.

Reply:
Hello @erdembayar

I can see that some of the tests failed in CI.
Would you be able to provide information which tests need my attention?

What else do you need me to do so that this can be merged?

Reply:
Thanks for the reply. I've rebased and squashed my change and pushed it to the branch.

Reply:
> Team Review: We think it would best if the code path that needs a custom comparison implements it that way rather than changing how all comparisons work. This could be a breaking change at the moment to alter how these objects are compared.

While I usually agree with being conservative with potential breaking changes, or introducing regressions, what we've seen in this PR with this one command needing a hack to work around a genuine problem means that other commands are likely to have the same bug. So, it's an engineering tradeoff.  Although, creating a new comparison class and in other PRs migrating those other commands to the new comparison might be an option. But if the current comparison ends up with nothing using it, that means it's buggy, so if we fix it, then we don't need to add new public APIs.

I haven't read this PR carefully enough, or know NuGet.Configuration's code well enough, to suggest my ideal solution. My opinion will certainly change as I learn more information. But given my limited knowledge right now, I think the original change to the comparison feels the best to me, unless someone can convince me otherwise.

> Several tests are failing, please check why they're failing: [#1](https://github.com/NuGet/NuGet.Client/blob/9836b4469a575b6b1aa7c9072c28b2f4ef6ec91d/test/NuGet.Core.Tests/NuGet.Configuration.Test/SettingsFileParsingTests/SourceItemTests.cs#L128), [#2](https://github.com/NuGet/NuGet.Client/blob/9836b4469a575b6b1aa7c9072c28b2f4ef6ec91d/test/NuGet.Core.Tests/NuGet.Configuration.Test/PackageSourceProviderTests.cs#L1029), [#3](https://github.com/NuGet/NuGet.Client/blob/9836b4469a575b6b1aa7c9072c28b2f4ef6ec91d/test/NuGet.Core.Tests/NuGet.Configuration.Test/PackageSourceProviderTests.cs#L1624).

I strongly believe all 3 tests are buggy, which means the product is buggy, and tests are asserting the bugs continue, rather than helping the bugs get fixed.

The first test is for the `SourceItem.Equals` method, which validates that a different key or a different protocol version are treated as not equal.  But why is `protocolVersion` being compared, whereas `value` (url/path) is not?  It makes so sense. Either `SourceItem.Equals` should validate all all properties (key, value, protocolVersion), or only `key`. The current implementation doesn't make sense.

The other two tests show the impact of this bug:

The second test assets that after "saving" two package sources, one of which has the same key as the test setup ("arrange"), that [`children[0]`](https://github.com/NuGet/NuGet.Client/blob/9836b4469a575b6b1aa7c9072c28b2f4ef6ec91d/test/NuGet.Core.Tests/NuGet.Configuration.Test/PackageSourceProviderTests.cs#L1059) and [`children[2]`](https://github.com/NuGet/NuGet.Client/blob/9836b4469a575b6b1aa7c9072c28b2f4ef6ec91d/test/NuGet.Core.Tests/NuGet.Configuration.Test/PackageSourceProviderTests.cs#L1063) have the same key. To me this really feels like a bug. When I hear `key` in a programming context, I think dictionary key, which needs to be unique. The test assertion shows that it's not unique. Why is this test asserting this?

Similarly, I find the third test very strange. In the test setup, it [pre-configures two sources with the same key](https://github.com/NuGet/NuGet.Client/blob/9836b4469a575b6b1aa7c9072c28b2f4ef6ec91d/test/NuGet.Core.Tests/NuGet.Configuration.Test/PackageSourceProviderTests.cs#L1633-L1634). Why? Like I commented on the previous test, I expect keys to be unique. The test also asserts that there are still two source items with the same key.

As described by our docs on NuGet configuration, NuGet "accumulates" settings from multiple files. If I have `c:\nuget.config` with `<add key="a" value="https://url1.test/" />`, and `c:\src\nuget.config` with `<add key="a" value="https://url2.test/" />`, then I'd expect the "accumulated" settings to have only one source with `key="a"`, and the rules that "closest wins", meaning "url2". The bug that this PR exposes is that if one of the two configs sets `protocolVersion="3"`, then the package source won't be updated, it'll be duplicated.

And within a single config file, there's no reason why two sources with the same key should be allowed to exist. In fact, for package source credentials, the source key is used as the XML element name, preventing the possibility of duplicates (but also forcing encoding of XML invalid characters).  If a nuget.config has two different package sources with the same key (but different protocolVersion, and different URL), and there's `packageSourceCredentials` for that key, which URL(s) should NuGet use the credentials for?

A different example, something I believe this PR is adding, is if I have a config file with 1 source that does not specify protocolVersion, and I run `dotnet nuget update source MySource --protocolVersion 3`, it would be very unexpected for a new source to be added with the same key/URL as the already existing source, right? What I'd expect as a customer is that `protocolVersion="3"` be added to the existing source in my config file.

I've edited this comment a few times as I've been writing it, so I can't remember if I've kept this my next comment in what I wrote above, but I believe this PR has exposed bugs in both the product and the tests. Normally it's not the place of of a PR author fixing adding a feature to fix a different bug, especially when the PR author is an external contributor. But in this case, the bug is directly related to the feature being added.

Reply:
coincidentally, a Microsoft-internal tester found a user-scenario where this bug (different protocolVersion considered not equal) causes a user-experience bug. I have a PR to fix that, and once it's merged, I can update this PR and get this one merged: https://github.com/NuGet/NuGet.Client/pull/5334

Reply:
Hello @zivkan,
Looking forward to #5334 being merged. After that I should be able to restart working on this PR.

Reply:
Updated code pushed.